cmd/manager/main.go

```diff
@@ -154,6 +153,14 @@ func main() {
 
 	log.Info("Starting the Cmd.")
 
+	go func() {
+		mgr.GetCache().WaitForCacheSync(stopChannel)
+		// create or update all NSTemplateTiers on the cluster at startup
+		if err := nstemplatetiers.CreateOrUpdateResources(mgr.GetScheme(), mgr.GetClient(), namespace, nstemplatetiers.Asset); err != nil {
+			log.Error(err, "")
+			os.Exit(1)
+		}
+	}()
```

Comment:
I'm a bit concerned that we run this in a separate go routine because of two reasons:
1. We create tiers asynchronously and I guess theoretically tiers might not be available when we need them. But probably our reconcile loops will handle it properly...
2. If something goes wrong then we exit here (but we don't know when it can happen). I'm not sure how all the controller update flow works... does it have any health check thing, etc so we can control rolling updates, etc.

But this is just something to keep in mind for now. I don't see how we could better.

Reply:
@alexeykazakov I understand your concerns. 

For 1., yes, our reconcile loop on the member-cluster operator should requeue the request if the NSTemplateTier resources are not available
For 2., the controller also waits for the cache to be in sync during its `Start` (see https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L140-L147), so I doubt we may face the situation where the go routine above kills the controller while it is already processing a request.

Reply:
> would it be too naive to wait for the cache to have it in sync before creating the tiers and starting the controllers?

@MatousJobanek what do you mean?

Reply:
@MatousJobanek I thought that the controllers started in the last block of this `main()` func, ie:
```
if err := mgr.Start(stopChannel); err != nil {
	log.Error(err, "Manager exited non-zero")
	os.Exit(1)
}
```
so I'm not sure what you mean here. Also, the controllers don't start before the cache is sync'd, as we can see in https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L140-L158. IMO, the controllers will start and at the same time, the NSTemplateTier resources will be created. So yes, there might be a very tiny time window when the controllers are running but there is no NSTemplateTier resources available. But then that would be the responsibility of the member-operator's controllers to requeue the requests until the NSTemplateTier resources exist. Or am I missing something?

Reply:
> what do you mean?

I meant to not start a separated goroutine and do it in the main one - which would mean to wait until the cache is synced directly in the main function - then create the tiers & start the controllers

> I thought that the controllers started in the last block of this main()

you are right, sorry for the confusion. I mixed it with the `StartKubeFedClusterControllers` function, but you don't have to care about this one.

>  So yes, there might be a very tiny time window

yes, I was referring to this tiny window


Anyway, you are right that we don't need to deal with it. I'm just trying to optimize it when it's not needed. 
So, you can ignore my comments above ;-).

Reply:
> I meant to not start a separated goroutine and do it in the main one - which would mean to wait until the cache is synced directly in the main function - then create the tiers & start the controllers

Can't do that because otherwise we would create a dead lock :(